src/node_large_page.cc

```diff
@@ -0,0 +1,278 @@
+// This is also handling the case w where the first line is not the binary
+
+static struct text_region FindNodeTextRegion() {
+  std::ifstream ifs;
+  std::string map_line;
+  std::string permission;
+  std::string dev;
+  char dash;
+  int64_t  start, end, offset, inode;
+  struct text_region nregion;
+
+  nregion.found_text_region = false;
+
+  ifs.open("/proc/self/maps");
+  if (!ifs) {
+    fprintf(stderr, "Could not open /proc/self/maps\n");
+    return nregion;
+  }
+
+  std::string exename;
+  {
+      char selfexe[ PATH_MAX ];
+      ssize_t count = readlink("/proc/self/exe", selfexe, PATH_MAX);
+      exename = std::string(selfexe, count);
+  }
+
+  while (std::getline(ifs, map_line)) {
+    std::istringstream iss(map_line);
+    iss >> std::hex >> start;
+    iss >> dash;
+    iss >> std::hex >> end;
+    iss >> permission;
+    iss >> offset;
+    iss >> dev;
+    iss >> inode;
+    if (inode != 0) {
+      std::string pathname;
+      iss >> pathname;
+      if (pathname.compare(exename) == 0) {
+        if (permission.compare("r-xp") == 0) {
+          start = reinterpret_cast<uint64_t>(&__nodetext);
+          char* from = reinterpret_cast<char*>(hugepage_align_up(start));
+          char* to = reinterpret_cast<char*>(hugepage_align_down(end));
+
+          if (from < to) {
+            size_t size = to - from;
+            nregion.found_text_region = true;
+            nregion.from = from;
+            nregion.to = to;
+            nregion.total_hugepages = size / hps;
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  ifs.close();
+  return nregion;
+}
+
+static bool IsTransparentHugePagesEnabled() {
+  std::ifstream ifs;
+
+  ifs.open("/sys/kernel/mm/transparent_hugepage/enabled");
+  if (!ifs) {
+    fprintf(stderr, "Could not open file: " \
+                    "/sys/kernel/mm/transparent_hugepage/enabled\n");
+    return false;
+  }
+
+  std::string always, madvise, never;
+  if (ifs.is_open()) {
+    while (ifs >> always >> madvise >> never) {}
+  }
+
+  int ret_status = false;
+
+  if (always.compare("[always]") == 0)
+    ret_status = true;
+  else if (madvise.compare("[madvise]") == 0)
+    ret_status = true;
+  else if (never.compare("[never]") == 0)
+    ret_status = false;
+
+  ifs.close();
+  return ret_status;
+}
+
+// Moving the text region to large pages. We need to be very careful.
+// 1: This function itself should not be moved.
+// We use a gcc option to put it outside the ".text" section
+// 2: This function should not call any function(s) that might be moved.
+// a. map a new area and copy the original code there
+// b. mmap using the start address with MAP_FIXED so we get exactly
+//    the same virtual address
+// c. madvise with MADV_HUGE_PAGE
+// d. If successful copy the code there and unmap the original region
+int
+__attribute__((__section__(".lpstub")))
+__attribute__((__aligned__(2 * 1024 * 1024)))
+__attribute__((__noinline__))
+__attribute__((__optimize__("2")))
```

Comment:
Is there a reason to set a fixed optimization level here?

Reply:
I added this as I wanted gcc to use O2 where no inlining of calls and no optimization of memcpy within this function happens. I am particularly concerned about the memcpy after the madvise call. If it gets replaced with something that is mapped out by the mmap call, it will crash here.

Reply:
I get the reasoning for `noinline` (and maybe you want `flatten` as well?) – I’m just a bit worried that setting a fixed optimization level might be a bit brittle?